service/controller/v18/resource/ipam/create_test.go

```diff
@@ -16,153 +19,355 @@ func mustParseCIDR(val string) net.IPNet {
 	return *n
 }
 
-func Test_canonicalizeSubnets(t *testing.T) {
+func Test_selectRandomAZs_properties(t *testing.T) {
 	testCases := []struct {
-		name            string
-		network         net.IPNet
-		subnets         []net.IPNet
-		expectedSubnets []net.IPNet
+		name         string
+		azs          []string
+		n            int
+		errorMatcher func(error) bool
+	}{}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			r := Resource{
+				logger:            microloggertest.New(),
+				availabilityZones: tc.azs,
+			}
+
+			azs, err := r.selectRandomAZs(tc.n)
+
+			switch {
+			case err == nil && tc.errorMatcher == nil:
+				// correct; carry on
+			case err != nil && tc.errorMatcher == nil:
+				t.Fatalf("error == %#v, want nil", err)
+			case err == nil && tc.errorMatcher != nil:
+				t.Fatalf("error == nil, want non-nil")
+			case !tc.errorMatcher(err):
+				t.Fatalf("error == %#v, want matching", err)
+			}
+
+			if tc.errorMatcher != nil {
+				return
+			}
+
+			if tc.n != len(azs) {
+				t.Fatalf("got %d AZs in the first round, expected %d", len(azs), tc.n)
+			}
+		})
+	}
+}
+
+func Test_selectRandomAZs_random(t *testing.T) {
+	originalAZs := []string{"eu-west-1a", "eu-west-1b", "eu-west-1c"}
+	r := Resource{
+		logger:            microloggertest.New(),
+		availabilityZones: originalAZs,
+	}
+
+	nTestRounds := 25
```

Comment:
In my local tests this test passed every time on third round. Having still some extra space eliminates flapping.